ee/clickhouse/sql/events.py

```diff
@@ -98,7 +105,16 @@
     database=settings.CLICKHOUSE_DATABASE,
 )
 
-KAFKA_EVENTS_TABLE_JSON_SQL = lambda: EVENTS_TABLE_BASE_SQL.format(
+# we add the settings to prevent poison pills from stopping ingestion
+# kafka_skip_broken_messages is an int, not a boolean, so we explicitly set
+# the max block size to consume from kafka such that we skip _all_ broken messages
+# this is an added safety mechanism given we control payloads to this topic
+KAFKA_EVENTS_TABLE_JSON_SQL = lambda: (
+    EVENTS_TABLE_BASE_SQL
+    + """
+    SETTINGS kafka_max_block_size=65505, kafka_skip_broken_messages=65505
```

Comment:
I wouldn't set `kafka_max_block_size` since that already defaults to `65536`

Are you sure you want to set `kafka_skip_broken_messages` to `65505`? Seems like we might drop a ton of events if we accidentally start emitting bad JSON formatted events.

>kafka_skip_broken_messages — the number of errors to tolerate when parsing messages, configured on a table level

A message is one JSON payload in this case for `JSONEachRow` which we are using here.

https://altinity.com/blog/clickhouse-kafka-engine-faq

Reply:
Well so the max block size here is our current default on the cluster. This here doesn't actually change it. I have it here so that we're very explicit.

As for `kafka_skip_broken_messages` should just be a boolean (as recognized by people in the ch repo). If we want it to operate as a boolean, we need to allow skipping as many messages as there are in the block, hence the block size is set here too to be explicit.

I think it makes sense to skip every broken message, because if we don't the consumer will just stop and keep crashing. We will anyway lose events as we'll need to move the consumer forward in some artificial way.

If we skip them or not, the events are lost, CH simply can't ingest them. The only thing is that with one we crash and with the other we don't.

Now, should we get a sense somehow of the amount of events CH is dropping? Probably

Reply:
Looks pretty clear to me
https://github.com/ClickHouse/ClickHouse/blob/87574e8bf60887f9b7df1e13b6bbdc3020a549ea/src/Storages/Kafka/KafkaSettings.h#L31
https://github.com/ClickHouse/ClickHouse/blob/972357ebf63dfecdc8d3ec968b0e9bf321fabf54/src/Storages/Kafka/StorageKafka.cpp#L254

Definitely not a boolean.

What you are suggesting is a really bad pattern. We potentially could be dropping a ton of events with no easy way of recovery. We wouldn't even know which events were successfully consumed or not consumed. We would have no metric to even detect if they are not getting consumed. With JSON you are asking to have a really bad time at some point in the future with this config.

Reply:
C'mon James, I'm not stupid, I know it's not a boolean, I wrote this schema, don't need to link me to the source code for confirmation. 

I meant it probably was meant to operate as one, and was referring to [this comment by a CH maintainer](https://github.com/ClickHouse/ClickHouse/issues/10647#issuecomment-623404036) that I didn't have in hand when writing this.

My only point here is that without this setting we'll fail on the _nth_ (e.g. 100th) payload, ingestion will stop, we'll have to debug, and we won't do anything with that payload anyway. But sure, I'm happy to set this lower, but as for "could be dropping a ton of events with no easy way of recovery" it's the same thing with or without this setting.

Reply:
I didn't mean to suggest you were dumb @yakkomajuri! We all know you are not!! You are the 🧠 s behind a ton of how we process stuff here.

I was just trying to make the point that it wasn't boolean. Sorry I might have misread your initial comment 😬 I thought you were suggesting that was a boolean field. I also apologize if I came across as harsh in the comment. I definitely did not mean it to come across that way. I just feel strongly on getting the kafka ingest parameters correct here because the side effects could be silently very bad. Believe me, I trust in your abilities. I trust you. Just trying to have a debate on configs here. Of that I can assure you. I think I need to spend more time crafting my PR reviews instead of drive by reviewing...

The main point I wanted to make was that that we _should_ break if we start having malformed json beyond a certain threshold right? Imagine this scenario: We change something in the plugin server, or a plugin, that breaks 1-3% of all events coming in from a JSON deserialization perspective on ClickHouse. It might be low enough in volume that we don't notice for some time - lets say 4-5 days (if we are lucky, it could be months or never). In this scenario we will end up losing events silently which is the thing that keeps me up at night the most. We should have a system that fails fast vs a system that is lossy, but very stable.

Alternatively what I've seen people do with these kinds of JSON serialized topologies is JSON kafka topic -> CH table with `String` column that has the body of the JSON -> Deserialized JSON CH table. That way if there are issues with deserializing you can keep those payloads around in a DLQ style table on CH. Something we might want to do honestly.

Side note from the link you provided:
>That feature will be redesigned (i hope soon). It will just have an _error virtual column instead.

This sounds _amazing_.

Reply:
I think that's a valid point - better to break fast. That does make sense